### PR TITLE
[SPARK-31615][SQL] Pretty string output for sql method of RuntimeReplaceable expressions

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

The RuntimeReplaceable ones are runtime replaceable, thus, their original parameters are not going to be resolved to PrettyAttribute and remain debug style string if we directly implement their `sql` methods with their parameters' `sql` methods.

This PR is raised with suggestions by @maropu and @cloud-fan https://github.com/apache/spark/pull/28402/files#r417656589. In this PR, we re-implement the `sql` methods of  the RuntimeReplaceable ones with toPettySQL

### Why are the changes needed?

Consistency of schema output between RuntimeReplaceable expressions and normal ones.

For example, `date_format` vs `to_timestamp`, before this PR, they output differently

#### Before
```sql
select date_format(timestamp '2019-10-06', 'yyyy-MM-dd uuuu')
struct<date_format(TIMESTAMP '2019-10-06 00:00:00', yyyy-MM-dd uuuu):string>

select to_timestamp("2019-10-06S10:11:12.12345", "yyyy-MM-dd'S'HH:mm:ss.SSSSSS")
struct<to_timestamp('2019-10-06S10:11:12.12345', 'yyyy-MM-dd\'S\'HH:mm:ss.SSSSSS'):timestamp>
```
#### After

```sql
select date_format(timestamp '2019-10-06', 'yyyy-MM-dd uuuu')
struct<date_format(TIMESTAMP '2019-10-06 00:00:00', yyyy-MM-dd uuuu):string>



select to_timestamp("2019-10-06T10:11:12'12", "yyyy-MM-dd'T'HH:mm:ss''SSSS")

struct<to_timestamp(2019-10-06T10:11:12'12, yyyy-MM-dd'T'HH:mm:ss''SSSS):timestamp>

````

### Does this PR introduce _any_ user-facing change?

Yes, the schema output style changed for the runtime replaceable expressions as shown in the above example

### How was this patch tested?
regenerate all related tests

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 1]
```diff
@@ -1206,7 +1207,7 @@ case class DatetimeSub(
     interval: Expression,
     child: Expression) extends RuntimeReplaceable {
   override def toString: String = s"$start - $interval"
-  override def sql: String = s"${start.sql} - ${interval.sql}"
+  override def sql: String = s"${toPrettySQL(start)} - ${toPrettySQL(interval)}"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:1210]: This means `RuntimeReplaceable.sql` always use pretty sql, while other expressions only use pretty sql when the caller side needs to (by calling `toPrettySQL`).

Is it possible to make them consistent? Like make `start` and `interval` as `innerChildren` and handle them in `usePrettyExpression`.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:1210]: OK, let me see

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:1210]: OK, let me see

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 2]
```diff
@@ -323,6 +323,20 @@ trait RuntimeReplaceable extends UnaryExpression with Unevaluable {
   // two `RuntimeReplaceable` are considered to be semantically equal if their "child" expressions
   // are semantically equal.
   override lazy val canonicalized: Expression = child.canonicalized
+
+  override def innerChildren: Seq[Expression] = sys.error("RuntimeReplaceable must implement" +
+    " innerChildren with the original parameters")
+
+  protected val sqlStrSeparator: String = ", "
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:330]: We need this variable? It seems only `Extract` uses this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 3]
```diff
@@ -323,6 +323,20 @@ trait RuntimeReplaceable extends UnaryExpression with Unevaluable {
   // two `RuntimeReplaceable` are considered to be semantically equal if their "child" expressions
   // are semantically equal.
   override lazy val canonicalized: Expression = child.canonicalized
+
+  override def innerChildren: Seq[Expression] = sys.error("RuntimeReplaceable must implement" +
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:327]: Since compilers cannot catch the case where one forget to override this, how about it like this?
```
  // RuntimeReplaceable
  def exprsReplaced: Seq[Expression]
  override def innerChildren: Seq[Expression] = exprsReplaced

  // e.g., ParseToTimestamp
  override def exprsReplaced: Seq[Expression] = Seq(Some(left), format).flatten

```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 4]
```diff
@@ -323,6 +323,20 @@ trait RuntimeReplaceable extends UnaryExpression with Unevaluable {
...
+    " innerChildren with the original parameters")
+
+  protected val sqlStrSeparator: String = ", "
+
+  override def sql: String = RuntimeReplaceable.this.prettyName +
+    prettyChildren.map(_.sql).mkString("(", sqlStrSeparator, ")")
+
+  protected var prettyChildren: Seq[Expression] = innerChildren
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:335]: I think its better to user `var` where possible, so how about it like this?
```
  // RuntimeReplaceable
  def newInstance(innerChildren: Seq[Expression]): RuntimeReplaceable

  // e.g., ParseToTimestamp
  override def newInstance(innerChildren: Seq[Expression]): ParseToTimestamp = {
    innerChildren match {
      case Seq(left) => ParseToTimestamp(left, None, child)
      case Seq(left, format) => ParseToTimestamp(left, Some(format), child)
    }
  }
```
Then, update the `toPrettySQL` in `utils.package`;
```
  def toPrettySQL(e: Expression): String = e match {
    case r: RuntimeReplaceable =>
      r.newInstance(r.innerChildren.map(usePrettyExpression)).sql
    case _ => usePrettyExpression(e).sql
  }
```

**Review by maropu — COMMENTED** [review]
Thanks for the followup, @yaooqinn

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:327]: Good Point!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 4 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:335]: The `newInstance` seems not robust to make a clone with `Seq[Expression]`

I think we can just add  `def prettySQL: String` in RuntimeReplaceable for  `utils.package` to call, and override it wherever needed without creating new instances.

Seems a better and simpler idea that comes up based on your suggestions, WDYT?

**Comment by maropu** [comment]
I feel having the pretty functionality in `RuntimeReplaceable` is a bit intrusive. Dropping the idea above is okay, but I think we need a more general solution for getting pretty strings outside `RuntimeReplaceable`. WDYT? @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 4 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:335]: I feel having the pretty functionality in `RuntimeReplaceable` is a bit intrusive. Dropping the idea above is okay, but I think we need a more general solution for getting pretty strings outside `RuntimeReplaceable`. WDYT? @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 5]
```diff
@@ -323,6 +324,19 @@ trait RuntimeReplaceable extends UnaryExpression with Unevaluable {
...
   override lazy val canonicalized: Expression = child.canonicalized
+
+  /**
+   * Only used to generate SQL representation of this expression.
+   *
+   * Implementations should override this with original parameters
+   */
+  def exprsReplaced: Seq[Expression]
+
+  override def sql: String = RuntimeReplaceable.this.prettyName +
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:335]: nit: seems we can remove `RuntimeReplaceable.this.`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 6]
```diff
@@ -2221,8 +2213,11 @@ case class Extract(field: Expression, source: Expression, child: Expression)
   }
 
   override def flatArguments: Iterator[Any] = Iterator(field, source)
-  override def sql: String = s"$prettyName(${field.sql} FROM ${source.sql})"
-  override def prettyName: String = "extract"
+  override def exprsReplaced: Seq[Expression] = Seq(field, source)
+
+  override def sql: String = super.sql.replace(", ", " FROM ")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:2218]: this is tricky and hard to read, can we just generate the sql string?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala [hunk 7]
```diff
@@ -142,7 +142,10 @@ package object util extends Logging {
     "`" + name.replace("`", "``") + "`"
   }
 
-  def toPrettySQL(e: Expression): String = usePrettyExpression(e).sql
+  def toPrettySQL(e: Expression): String = e match {
+    case r: RuntimeReplaceable => r.prettySQL
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:146]: does this work if `RuntimeReplaceable` is not the root node?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:146]: Ah.. no..

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala]: https://github.com/apache/spark/pull/28420/files#diff-6f4edc80e2cc973e748705e85a6053b4R788-R807

fixed and added some nested tests, thanks, good catch!

**Comment by yaooqinn** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 8]
```diff
@@ -323,6 +324,20 @@ trait RuntimeReplaceable extends UnaryExpression with Unevaluable {
...
+   * Implementations should override this with original parameters
+   */
+  protected def exprsReplaced: Seq[Expression]
+
+  override def sql: String = prettyName + exprsReplaced.map(_.sql).mkString("(", ", ", ")")
+
+  protected def prettySQL: String =
+    prettyName + exprsReplaced.map(toPrettySQL).mkString("(", ", ", ")")
+
+  def asPrettyAttribute(): PrettyAttribute = PrettyAttribute(prettySQL, dataType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:340]: why not add a `def withPrettyAttributes: this.type`? The current way requires fewer changes as we can have a default implementation, but it's hacky to convert `RuntimeReplaceable` to a single `PrettyAttribute`.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: I think it might be better to move this to `util.package`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 9]
```diff
@@ -323,6 +324,18 @@ trait RuntimeReplaceable extends UnaryExpression with Unevaluable {
...
   override lazy val canonicalized: Expression = child.canonicalized
+
+  /**
+   * Only used to generate SQL representation of this expression.
+   *
+   * Implementations should override this with original parameters
+   */
+  protected def exprsReplaced: Seq[Expression]
+
+  override def sql: String = prettyName + exprsReplaced.map(_.sql).mkString("(", ", ", ")")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:335]: another idea: maybe we can define a `mkString` method?
```
def exprsReplaced: Seq[Expression]

def mkString(childrenString: Seq[String]): String = childrenString.mkString("(", ", ", ")")

override def sql: String = prettyName + mkString(exprsReplaced.map(_.sql))
```

then in `util`
```
case r: RuntimeReplaceable => PrettyAttribute(r.mkString(r.exprsReplaced.map(toPrettySQL)), r.dataType)
```

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:335]: SGTM, this way is much better.  We should put the `prettyName` in `mkString` though.

**Review by cloud-fan — APPROVED** [review]

**Comment by yaooqinn** [comment]
retest this please

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
We need to backport this into branch-2.4, too? I personally think printing debug info looks like a bug. @cloud-fan @dongjoon-hyun

**Comment by cloud-fan** [comment]
It's a minor bug, but the diff is not small. Maybe master-only is good enough?

**Comment by maropu** [comment]
ok

**Comment by maropu** [comment]
Thanks! Merged to master.

File: sql/core/src/test/resources/sql-tests/results/predicate-functions.sql.out [hunk 10]
```diff
@@ -245,54 +245,54 @@ true
 -- !query
 select to_date('2009-07-30 04:17:52') <= to_date('2009-07-30 04:17:52')
 -- !query schema
-struct<(to_date('2009-07-30 04:17:52') <= to_date('2009-07-30 04:17:52')):boolean>
+struct<(to_date(2009-07-30 04:17:52) <= to_date(2009-07-30 04:17:52)):boolean>
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/results/predicate-functions.sql.out:248]: Sorry I happened to leave a comment late but why was this changed? Seems the previous one looks better.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 11]
```diff
@@ -323,6 +324,19 @@ trait RuntimeReplaceable extends UnaryExpression with Unevaluable {
...
+   *
+   * Implementations should override this with original parameters
+   */
+  def exprsReplaced: Seq[Expression]
+
+  override def sql: String = mkString(exprsReplaced.map(_.sql))
+
+  def mkString(childrenString: Seq[String]): String = {
+    prettyName + childrenString.mkString("(", ", ", ")")
+  }
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:339]: Why do we need two extra interface methods just to make the column name pretty? Seems the output isn't even completely pretty.

File: sql/core/src/test/resources/sql-tests/results/predicate-functions.sql.out (hunk 10 above)
Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/results/predicate-functions.sql.out:248]: check this
```sql
+-- !query
+select date_format('2019-10-06', 'yyyy-MM-dd uuuu')
+-- !query schema
+struct<date_format(CAST(2019-10-06 AS TIMESTAMP), yyyy-MM-dd uuuu):string>
+-- !query output
+2019-10-06 Sunday
```

Comment (HyukjinKwon) [reply sql/core/src/test/resources/sql-tests/results/predicate-functions.sql.out:248]: Did we discuss about changing ^ case instead?

Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/results/predicate-functions.sql.out:248]: Not yet. This PR just makes the `RuntimeReplaceable` -> `PrettyAttribute`. I agree with you that the `PrettyAttribute` is not beautiful in itself.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 11 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:339]: It's not about pretty or not. It's about consistency. If we call a function with its alias name, then it will not use `PrettyAttribute` and the string representation is different from if we call the function with the original name.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:339]: The consistency is good but the readability here is the problem. See the implementation side:

```scala
  override def sql: String = s"$prettyName(${field.sql} FROM ${source.sql})"
  override def prettyName: String = "extract"
```
vs
```scala
  override def exprsReplaced: Seq[Expression] = Seq(field, source)

  override def mkString(childrenString: Seq[String]): String = {
    prettyName + childrenString.mkString("(", " FROM ", ")")
  }
```

Previously, it was fine to only know `sql` and `prettyName` but now you should know what `exprsReplaced` and `mkString` are. From the naming, it's confusing what that means too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala [hunk 12]
```diff
@@ -134,6 +134,8 @@ package object util extends Logging {
       PrettyAttribute(usePrettyExpression(e.child).sql + "." + name, e.dataType)
     case e: GetArrayStructFields =>
       PrettyAttribute(usePrettyExpression(e.child) + "." + e.field.name, e.dataType)
+    case r: RuntimeReplaceable =>
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:137]: And strictly `RuntimeReplaceable` itself isn't an attribute ...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 11 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:339]: I would just document that `sql` and `prettyName` should be properly overridden at `RuntimeReplaceable` just like we have the restriction about the `child`.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:339]: The caller side decides if he wants pretty string or not, by calling `Expression.sql` or `toPrettySQL(Expression)`. So simply overridden `sql` doesn't work, unless we add a boolean parameter.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala (hunk 12 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:137]: We already did the same to `GetStructField`  :)
